### PR TITLE
planner: Add trace for agg pushdown rule

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/29661

### What is changed and how it works?

Add optimize trace step in rule aggregation push down.

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
None
```

### CODE REVIEW

File: planner/core/logical_plan_trace_test.go [hunk 1]
```diff
@@ -101,6 +101,36 @@ func (s *testPlanSuite) TestSingleRuleTraceStep(c *C) {
 				},
 			},
 		},
+		{
+			sql:            "select count(*) from t a , t b, t c",
+			flags:          []uint64{flagBuildKeyInfo, flagPrunColumns, flagPushDownAgg},
+			assertRuleName: "aggregation_push_down",
+			assertRuleSteps: []assertTraceStep{
+				{
+					assertAction: "aggFuncs[count(Column#38)] pushed into join[5] right path",
```
Comment (rebelice) [inline planner/core/logical_plan_trace_test.go:110]: Just comment. Is it better to add the expression here?

**Review by rebelice — APPROVED** [review]

File: planner/core/logical_plan_trace_test.go (hunk 1 above)
Comment (Yisaer) [reply planner/core/logical_plan_trace_test.go:121]: Currently not necessary.

File: planner/core/rule_aggregation_push_down.go [hunk 2]
```diff
@@ -500,3 +513,70 @@ func (a *aggregationPushDownSolver) aggPushDown(p LogicalPlan, opt *logicalOptim
...
+		buffer := bytes.NewBufferString("agg[%v]'s functions[")
+		for i, aggFunc := range aggFuncs {
+			if i > 0 {
+				buffer.WriteString(",")
+			}
+			buffer.WriteString(aggFunc.String())
+		}
+		buffer.WriteString("] are decomposable with join")
+		return buffer.String()
+	}()
```
Comment (time-and-fate) [inline planner/core/rule_aggregation_push_down.go:529]: Why do we use functions here and below?

File: planner/core/rule_aggregation_push_down.go [hunk 3]
```diff
@@ -500,3 +513,70 @@ func (a *aggregationPushDownSolver) aggPushDown(p LogicalPlan, opt *logicalOptim
 func (*aggregationPushDownSolver) name() string {
 	return "aggregation_push_down"
 }
+
+func appendAggPushDownAcrossJoinTraceStep(oldAgg, newAgg *LogicalAggregation, aggFuncs []*aggregation.AggFuncDesc, join *LogicalJoin,
+	childIdx int, opt *logicalOptimizeOp) {
+	reason := func() string {
+		buffer := bytes.NewBufferString("agg[%v]'s functions[")
```
Comment (time-and-fate) [inline planner/core/rule_aggregation_push_down.go:520]: Missing argument for `%v`.

File: planner/core/rule_aggregation_push_down.go [hunk 4]
```diff
@@ -500,3 +513,70 @@ func (a *aggregationPushDownSolver) aggPushDown(p LogicalPlan, opt *logicalOptim
...
+		for i, aggFunc := range agg.AggFuncs {
+			if i > 0 {
+				buffer.WriteString(",")
+			}
+			buffer.WriteString(aggFunc.String())
+		}
+		buffer.WriteString("]")
+		return buffer.String()
+	}()
+	opt.appendStepToCurrent(agg.ID(), agg.TP(), "projection can be eliminated", action)
```
Comment (time-and-fate) [inline planner/core/rule_aggregation_push_down.go:555]:
```suggestion
	opt.appendStepToCurrent(agg.ID(), agg.TP(), "projection is directly below an aggregation and has no side effects", action)
```

File: planner/core/rule_aggregation_push_down.go (hunk 2 above)
Comment (Yisaer) [reply planner/core/rule_aggregation_push_down.go]: To build reason/action in a separate block.

Comment (time-and-fate) [reply planner/core/rule_aggregation_push_down.go]: Ok...though I think we don't need a function to achieve this.

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**